### ISSUE TITLE
Add Specific Ability for Player Flagship to Dock with Escorts

### ISSUE THREAD

**Issue by QuantumTux** [endless-sky/endless-sky#9707]
### Problem Description

This is a problem; it is partially related to plundering/capturing, but does go beyond that to basic game playability and enjoyment. It's also a very old problem. [^1]

The player shouldn't have to "drop" cargo and **_hope_** that an Escort can pick it up before someone else grabs it. This is especially a problem when the player's fleet is being swarmed by NPC units scanning them (which is a [separate, annoying problem (Issue #9706)](https://github.com/endless-sky/endless-sky/issues/9706)). [^2]

Solving this problem opens the door to interesting capabilities.

### Related Issue Links

[Issue #651](https://github.com/endless-sky/endless-sky/issues/651) - When boarding friendly ships, you should have the option to help or attack
[Issue #5338](https://github.com/endless-sky/endless-sky/issues/5338) - feat: New Dock-command and not-boarder keyword
[Issue #5709](https://github.com/endless-sky/endless-sky/issues/5709) - Idea: Possible solution to repeat boarding/capping
[Issue #6323](https://github.com/endless-sky/endless-sky/issues/6323) - feat(in-flight-docking): Allow fighter-flagships to dock with carrier during flight
[Issue #7030](https://github.com/endless-sky/endless-sky/issues/7030) - fix(flagship fighters): Proper carry and handling for ex-flagship fighters
[Issue #9218](https://github.com/endless-sky/endless-sky/issues/9218) - enhancement: order escorts to plunder ships

### Desired Solution

"Docking" should be recognized as a separate operation/action, distinct from "Boarding":

- Docking: A player ship links to another player ship **under the player's control** (thus, a player can only **board** a **_disabled_** Escort)
- Boarding: A ship links to another ship, where one ship is **not under the player's control**

The player flagship should be able to **_dock_** with a selected Escort, and a key binding should be assigned for this purpose.

While Docked, the two ships may:

- Exchange crew (however, no ship can be reduced below minimum crew, or exceed the number of available bunks)
- Transfer cargo **without** having to "drop" cargo and _hope_ the other ship can pick it up (however, sufficient Cargo space must exist to receive any transfers)
- Perform repairs; if one ship is damaged and the other ship is both undamaged and has Repair capability, then **_a portion_** [^3] of the Repair generated by the Repair-capable ship may be applied to the damaged ship (see **Additional Context**)
- Transfer power; if one ship has power needs in excess of its generation capability, and the other ship has excess power generation, then **_a portion_** [^4] of the excess power generation may be transferred to the other ship [^5] (see **Additional Context**)

Docking does **not** permit:

- Shield regeneration for a ship not equipped with a shield generator
- Installation or removal of Outfits
- Movement while docked

### Alternative Approaches

I suppose that it would be possible to use the same masking that determines whether or not, for example, a missile hits to also force which ship can pick up something dropped by a friendly "touching" ship.

### Additional Context

As I said, if "Docking" (as described) is used to solve the original problem, then it opens the door to adding some new capabilities. I'm referring to some previous **Related Issues** specific to my proposals

- [Issue #7333](https://github.com/endless-sky/endless-sky/issues/7333) - Outfit durability and repairs

The suggested "Repair transfer" capability addresses a problem tangential to the original problem. When an Escort gets disabled, it can be "repaired" enough to get it moving, but it's still a flying eggshell. If the Escort isn't equipped with a Repair Outfit, there's no way to do anything else for it; when involved in detached operations, that can be fatal.

This proposal gives a way for a player to repair a damaged ship without requiring a Repair Outfit on everything. The trade-off is that the repairs happen very slowly.

Regarding the suggsted "Power transfer" capability, the main use that comes to mind is aiding the _speed_ of shield regeneration. Of all my suggestions in this Issue, this is the one I'm least-certain is workable/useful.

[^1]: This has been a problem for as long as I've been playing ES; and when I started, the Korath only had one type of Heat Shunt
[^2]: I plan to open a separate Issue regarding this
[^3]: The limit should be fairly low; for example, if Ship B is damaged and Ship A is intact **and** generates, say, 20 Repair, then perhaps only 5 Repair should be applicable to Ship B
[^4]: Again, this should be very limited; for example, if Ship B is short on power, then Ship A can transfer probably no more than 10% of its generation to Ship B
[^5]: No matter what, the donating ship cannot enter into power deficit

**Comment by Hurleveur** [comment]
I dont think we should allow transfering crew as all it'd do is make boarding easier by allowing you to stack crew on non flagship ships and get it back

Its probably best to keep it simple and limited to transfering cargo to begin with - I wouldnt mind a boost to looting

**Comment by QuantumTux** [comment]
@Hurleveur If you can transfer Cargo, I'm hard pressed to think of a reason it wouldn't be possible to transfer crew.

Yes, this proposal **would** make it easier to conduct multiple captures, or go after a bigger target, without having to land. It would be possible to attack a very big target, whittle down the defenders a bit, Dock to an Escort and replace crew combat losses, re-board, whittle down the defenders a bit more, Dock to another Escort, replace crew combat losses, rinse and repeat until you run out of spare crew or finally get the prize. One could run up quite the butcher's bill, but in normal gameplay, Pirates are slaughtered like dogs for months on end - it's not like one can finish most of the ES campaigns without racking up enough dead bodies to make even John Wick wince.

**Comment by Hurleveur** [comment]
I mean you just said the reason yourself, and boarding is op enough as it is

Realism does not matter, it's more about gameplay and lore

**Comment by xX-Dillinger-Xx** [comment]
I really like this idea. I agree it could add some interesting possibilities.
Boarding and docking should be separate commands. As it stands, "B" auto selects non owned ship when sometimes what I want to do is rescue my escorts first, before more hostiles show up. A '_Board'_ hotkey could select hostile ships and a separate '_Dock_' hotkey could select owned and non owned escorts. Transferring cargo while 'docked' is a great idea.
As for crew transfer, I think I remember reading that they may be adding some limits on boarding. Limiting it to 1 attempt would cut down on the OP aspect and still allow you transfer crew before you try to capture. I have always found it frustrating that the one time you finally disable that rare ship, you've been after, you don't have all your crew onboard to capture it. This usually happens if you just finished capturing something else and then, like magic, that rare jewel shows up. If you land to re-hire then that rare prize is gone. 
I'm not sure about everybody else but I usually have to choose which flag ship I'm going to pilot based off what I'm planning next. Am I planning on trying to plunder outfits or capture? So, maybe another option might be to allow you to transfer yourself to another flagship while 'docked'. To limit the OP of this option the player could have to preselect Flagship A and Flagship B before taking off and then be allowed to transfer yourself between "A" and "B" only.

**Comment by xX-Dillinger-Xx** [comment]
I just had another thought. As the commander of my fleet, I should be able to issue orders. "Captain of escort '_xyz'_ go board that disabled ship, tell me what's on board and I'll tell you what to take, or whether to attempt capture." Obviously, all based off of "xyz's" cargo space and crew compliment. Also, the player would  just be presented the already existing dialogue box.

**Comment by junetried** [comment]
One problem with docking and repairing escorts is that it doesn't make sense in all contexts - it sounds feasible that the microbots from a Systems Core or Repair Module could be transferred to another ship, but how does the hull of a Remnant ship regenerate another ship's hull?

**Comment by xX-Dillinger-Xx** [comment]
> ....the whole problem of how over powered and easily abused the salvage & capture system is currently. There are efforts in the works to solve this, but I'm not sure how close any of the actual solutions are to being implemented. This is something that a few tweaks to H2H values isn't going to fix.

I agree that the current system needs a rework. I'm aware that most ideas that get suggested require significant work to rewrite and implement. At least, here dev's are open to suggestions and willing to discuss others points of view. Thanks for that. Discussion of ideas and suggestions is very important because most implemented changes and features end up being an amalgamation of many peoples input.  
Years ago I spent a lot of time playing a "free to play - pay to win" MMO game and was also heavily involved in their forums. The problem there was, that they rarely listened to anything the players had to say. Their only objective was implementing more ways for players to pay and game quality suffered. The fun factor suffered even worse. Players started leaving and now that game, only has a fraction of the players it once did. IMO greed and the dev's egos destroyed that game.

**Comment by Mernom** [comment]
A way to nerf crew transfer boarding cheese is to add a penalty to attack power for every transfered crew member, due to moral loss. 
IMO crew transfer is some form is very vital, since I've had many cases where I managed to capture a valuable ship, and then had to spend 10 minutes trying to land on the planet becasue my ship kept malfunctioning right when it was supposed to counter thrust for landing.

Here are some more ideas for limitations: The donating ship can't go below 80% of it's crew, the recieving ship can't go above 100% of it's crew, moral penalty for attacking proportional to how many of your crew is donated.

Also, some sort of way to hire reserves for your ships to facilitate this transfer would be needed.

BTW, pro tip for cargo transfer by dumping it: order your cargo escorts to all overlap in one spot, overlap with them, and _then_ dump.
Did that to great effect in Ka'het and Korath territory in my latest playthrough.
